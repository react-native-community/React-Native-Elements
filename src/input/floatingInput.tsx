import React from 'react';
import {
  View,
  Animated,
  StyleSheet,
  TextInput,
  TextStyle,
  StyleProp,
  TextInputProps,
} from 'react-native';
import { RneFunctionComponent } from '../helpers';
import { withTheme } from '../config';
export type FloatingInputProps = TextInputProps & {
  label?: string;
  value?: string;
  update?: (attrName: string, updatedValue: string) => any;
  keyboardType?: string;
  labelActiveSize?: number;
  labelInActiveSize?: number;
  labelActiveColor?: string;
  labelInActiveColor?: string;
  textInputStyles?: StyleProp<TextStyle>;
  otherTextInputProps?: {};
};
const FloatingTextInput: RneFunctionComponent<FloatingInputProps> = ({
  keyboardType = 'default',
  labelActiveSize = 14,
  labelInActiveSize = 15,
  labelActiveColor = '#86939E',
  labelInActiveColor = '#5E6977',
  textInputStyles = {},
  ...props
}) => {
  const [isFieldActive, setIsFieldActive] = React.useState(false);
  const { current: position } = React.useRef(
    new Animated.Value(isFieldActive ? 1 : 0)
  );
  const handleFocus = () => {
    if (!isFieldActive) {
      setIsFieldActive(true);
      Animated.timing(position, {
        toValue: 1,
        duration: 250,
        useNativeDriver: false,
      }).start();
    }
  };
  const handleBlur = () => {
    if (isFieldActive && !props.value) {
      setIsFieldActive(false);
      Animated.timing(position, {
        toValue: 0,
        duration: 250,
        useNativeDriver: false,
      }).start();
    }
  };

  const returnAnimatedlabelStyle = () => {
    return {
      top: position.interpolate({
        inputRange: [0, 1],
        outputRange: [14, 0],
      }),
      fontSize: isFieldActive ? labelActiveSize : labelInActiveSize,
      color: isFieldActive ? labelActiveColor : labelInActiveColor,
    };
  };
  return (
    <View style={styles.container}>
      <Animated.Text style={[styles.labelStyles, returnAnimatedlabelStyle()]}>
        {props.label}
      </Animated.Text>
      <TextInput
        style={[styles.textInput, textInputStyles]}
        value={props.value}
        underlineColorAndroid="transparent"
        onFocus={handleFocus}
        onBlur={handleBlur}
        keyboardType={keyboardType}
        {...props.otherTextInputProps}
      />
    </View>
  );
};
const styles = StyleSheet.create({
  container: {
    width: '90%',
    borderRadius: 5,
    borderStyle: 'solid',
    borderColor: '#86939E',
    borderWidth: 0.5,
    height: 50,
    marginVertical: 6,
  },
  textInput: {
    fontSize: 15,
    marginTop: 20,
    marginLeft: 13,
    color: '#86939E',
  },
  labelStyles: {
    position: 'absolute',
    left: 13,
  },
});
export { FloatingTextInput };
export default withTheme(FloatingTextInput, 'FloatingInput');
