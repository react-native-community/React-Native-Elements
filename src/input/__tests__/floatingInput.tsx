import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import theme from '../../config/theme';
import { FloatingTextInput } from '../floatingInput';
import { Input } from '../Input';

describe('Input component', () => {
  beforeAll(() => {
    // useNativeDriver isn't available in jest, so just silencing the warning
    global.console.warn = () => null;
  });

  it('should match snapshot', () => {
    const component = shallow(<FloatingTextInput theme={theme} />);
    expect(component.length).toBe(1);
    expect(toJson(component)).toMatchSnapshot();
  });

  describe('Props', () => {
    it('label', () => {
      const component = shallow(
        <FloatingTextInput theme={theme} label="My Label Name" />
      );
      expect(component.length).toBe(1);
      expect(toJson(component)).toMatchSnapshot();
    });

    it('value', () => {
      const component = shallow(
        <FloatingTextInput theme={theme} value="My Value" />
      );
      expect(component.length).toBe(1);
      expect(toJson(component)).toMatchSnapshot();
    });

    it('labelActiveSize', () => {
      const component = shallow(<FloatingTextInput labelActiveSize={14} />);
      expect(component.length).toBe(1);
      expect(toJson(component)).toMatchSnapshot();
    });

    it('labelInActiveSize', () => {
      const component = shallow(<FloatingTextInput labelInActiveSize={15} />);
      expect(component.length).toBe(1);
      expect(toJson(component)).toMatchSnapshot();
    });

    it('labelActiveColor', () => {
      const component = shallow(
        <FloatingTextInput theme={theme} labelActiveColor="#86939E" />
      );
      expect(component.length).toBe(1);
      expect(toJson(component)).toMatchSnapshot();
    });

    it('labelInActiveColor', () => {
      const component = shallow(
        <FloatingTextInput theme={theme} labelInActiveColor="#5E6977" />
      );
      expect(component.length).toBe(1);
      expect(toJson(component)).toMatchSnapshot();
    });
  });

  describe('Instance methods', () => {
    it('should focus the input', () => {
      const focus = jest.fn();
      const component = shallow(<Input theme={theme} />);
      const instance = component.instance();
      // Refs not available in shallow render
      instance.input = {
        focus,
      };
      instance.focus();
      expect(focus).toHaveBeenCalledTimes(1);
    });

    it('should blur the input', () => {
      const blur = jest.fn();
      const component = shallow(<Input theme={theme} />);
      const instance = component.instance();
      // Refs not available in shallow render
      instance.input = {
        blur,
      };
      instance.blur();
      expect(blur).toHaveBeenCalledTimes(1);
    });
  });
});
