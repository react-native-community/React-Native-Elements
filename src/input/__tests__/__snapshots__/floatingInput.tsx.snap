// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input component Props label 1`] = `
<View
  style={
    Object {
      "borderColor": "#86939E",
      "borderRadius": 5,
      "borderStyle": "solid",
      "borderWidth": 0.5,
      "height": 50,
      "marginVertical": 6,
      "width": "90%",
    }
  }
>
  <ForwardRef(AnimatedComponentWrapper)
    style={
      Array [
        Object {
          "left": 13,
          "position": "absolute",
        },
        Object {
          "color": "#5E6977",
          "fontSize": 15,
          "top": 14,
        },
      ]
    }
  >
    My Label Name
  </ForwardRef(AnimatedComponentWrapper)>
  <Component
    allowFontScaling={true}
    keyboardType="default"
    onBlur={[Function]}
    onFocus={[Function]}
    rejectResponderTermination={true}
    style={
      Array [
        Object {
          "color": "#86939E",
          "fontSize": 15,
          "marginLeft": 13,
          "marginTop": 20,
        },
        Object {},
      ]
    }
    underlineColorAndroid="transparent"
  />
</View>
`;

exports[`Input component Props labelActiveColor 1`] = `
<View
  style={
    Object {
      "borderColor": "#86939E",
      "borderRadius": 5,
      "borderStyle": "solid",
      "borderWidth": 0.5,
      "height": 50,
      "marginVertical": 6,
      "width": "90%",
    }
  }
>
  <ForwardRef(AnimatedComponentWrapper)
    style={
      Array [
        Object {
          "left": 13,
          "position": "absolute",
        },
        Object {
          "color": "#5E6977",
          "fontSize": 15,
          "top": 14,
        },
      ]
    }
  />
  <Component
    allowFontScaling={true}
    keyboardType="default"
    onBlur={[Function]}
    onFocus={[Function]}
    rejectResponderTermination={true}
    style={
      Array [
        Object {
          "color": "#86939E",
          "fontSize": 15,
          "marginLeft": 13,
          "marginTop": 20,
        },
        Object {},
      ]
    }
    underlineColorAndroid="transparent"
  />
</View>
`;

exports[`Input component Props labelActiveSize 1`] = `
<View
  style={
    Object {
      "borderColor": "#86939E",
      "borderRadius": 5,
      "borderStyle": "solid",
      "borderWidth": 0.5,
      "height": 50,
      "marginVertical": 6,
      "width": "90%",
    }
  }
>
  <ForwardRef(AnimatedComponentWrapper)
    style={
      Array [
        Object {
          "left": 13,
          "position": "absolute",
        },
        Object {
          "color": "#5E6977",
          "fontSize": 15,
          "top": 14,
        },
      ]
    }
  />
  <Component
    allowFontScaling={true}
    keyboardType="default"
    onBlur={[Function]}
    onFocus={[Function]}
    rejectResponderTermination={true}
    style={
      Array [
        Object {
          "color": "#86939E",
          "fontSize": 15,
          "marginLeft": 13,
          "marginTop": 20,
        },
        Object {},
      ]
    }
    underlineColorAndroid="transparent"
  />
</View>
`;

exports[`Input component Props labelInActiveColor 1`] = `
<View
  style={
    Object {
      "borderColor": "#86939E",
      "borderRadius": 5,
      "borderStyle": "solid",
      "borderWidth": 0.5,
      "height": 50,
      "marginVertical": 6,
      "width": "90%",
    }
  }
>
  <ForwardRef(AnimatedComponentWrapper)
    style={
      Array [
        Object {
          "left": 13,
          "position": "absolute",
        },
        Object {
          "color": "#5E6977",
          "fontSize": 15,
          "top": 14,
        },
      ]
    }
  />
  <Component
    allowFontScaling={true}
    keyboardType="default"
    onBlur={[Function]}
    onFocus={[Function]}
    rejectResponderTermination={true}
    style={
      Array [
        Object {
          "color": "#86939E",
          "fontSize": 15,
          "marginLeft": 13,
          "marginTop": 20,
        },
        Object {},
      ]
    }
    underlineColorAndroid="transparent"
  />
</View>
`;

exports[`Input component Props labelInActiveSize 1`] = `
<View
  style={
    Object {
      "borderColor": "#86939E",
      "borderRadius": 5,
      "borderStyle": "solid",
      "borderWidth": 0.5,
      "height": 50,
      "marginVertical": 6,
      "width": "90%",
    }
  }
>
  <ForwardRef(AnimatedComponentWrapper)
    style={
      Array [
        Object {
          "left": 13,
          "position": "absolute",
        },
        Object {
          "color": "#5E6977",
          "fontSize": 15,
          "top": 14,
        },
      ]
    }
  />
  <Component
    allowFontScaling={true}
    keyboardType="default"
    onBlur={[Function]}
    onFocus={[Function]}
    rejectResponderTermination={true}
    style={
      Array [
        Object {
          "color": "#86939E",
          "fontSize": 15,
          "marginLeft": 13,
          "marginTop": 20,
        },
        Object {},
      ]
    }
    underlineColorAndroid="transparent"
  />
</View>
`;

exports[`Input component Props value 1`] = `
<View
  style={
    Object {
      "borderColor": "#86939E",
      "borderRadius": 5,
      "borderStyle": "solid",
      "borderWidth": 0.5,
      "height": 50,
      "marginVertical": 6,
      "width": "90%",
    }
  }
>
  <ForwardRef(AnimatedComponentWrapper)
    style={
      Array [
        Object {
          "left": 13,
          "position": "absolute",
        },
        Object {
          "color": "#5E6977",
          "fontSize": 15,
          "top": 14,
        },
      ]
    }
  />
  <Component
    allowFontScaling={true}
    keyboardType="default"
    onBlur={[Function]}
    onFocus={[Function]}
    rejectResponderTermination={true}
    style={
      Array [
        Object {
          "color": "#86939E",
          "fontSize": 15,
          "marginLeft": 13,
          "marginTop": 20,
        },
        Object {},
      ]
    }
    underlineColorAndroid="transparent"
    value="My Value"
  />
</View>
`;

exports[`Input component should match snapshot 1`] = `
<View
  style={
    Object {
      "borderColor": "#86939E",
      "borderRadius": 5,
      "borderStyle": "solid",
      "borderWidth": 0.5,
      "height": 50,
      "marginVertical": 6,
      "width": "90%",
    }
  }
>
  <ForwardRef(AnimatedComponentWrapper)
    style={
      Array [
        Object {
          "left": 13,
          "position": "absolute",
        },
        Object {
          "color": "#5E6977",
          "fontSize": 15,
          "top": 14,
        },
      ]
    }
  />
  <Component
    allowFontScaling={true}
    keyboardType="default"
    onBlur={[Function]}
    onFocus={[Function]}
    rejectResponderTermination={true}
    style={
      Array [
        Object {
          "color": "#86939E",
          "fontSize": 15,
          "marginLeft": 13,
          "marginTop": 20,
        },
        Object {},
      ]
    }
    underlineColorAndroid="transparent"
  />
</View>
`;
